perl-lib/OESS/lib/OESS/MPLS/Device/Juniper/MX.pm

```diff
@@ -1641,62 +1641,70 @@ sub _is_active_circuit{
 
 =head2 get_device_diff
 
+    my ($diff, $err) = get_device_diff($xml_configuration);
+    if ($err) {
+        $logger->error($err);
+    }
+
 Returns and stores a human readable diff for display to users.
 
 =cut
 sub get_device_diff {
     my $self = shift;
     my $conf = shift;
 
-    if(!$self->connected()){
-        $self->{'logger'}->error("Not currently connected to device");
-        return;
+    if (!$self->connected()) {
+        return (undef, "Not currently connected to device");
     }
 
     $self->{'logger'}->debug("Candidate config: " . $conf);
 
-    my %queryargs = ('target' => 'candidate');
-    $queryargs{'config'} = $conf;
+    my %queryargs = (
+        'config' => $conf,
+        'target' => 'candidate',
+    );
 
     my $ok = $self->lock();
     if (!$ok) {
-        $self->{'logger'}->error("Unable to generate diff without a valid lock.");
-        return;
+        return (undef, "Unable to generate diff without a valid lock.");
     }
 
     my $res = $self->{'jnx'}->edit_config(%queryargs);
     if ($self->{'jnx'}->has_error) {
-	my $dom = $self->{'jnx'}->get_dom();
-	$self->{'logger'}->error($dom->toString());
+        my $dom = $self->{'jnx'}->get_dom();
+        $self->{'logger'}->error($dom->toString());
+
         my $error = $self->{'jnx'}->get_first_error();
         if ($error->{'error_message'} !~ /uncommitted/) {
-            $self->{'logger'}->error("Error getting device diff: " . $error->{'error_message'});
             $self->disconnect();
-            return;
+            return (undef, "Error getting device diff: $error->{error_message}");
         }
     }
 
     # According to docs format isn't considered when used with
     # compare. However in 15.1F6-S6.4 it is; I would expect this to
     # continue in later Junos versions.
-    my $configcompare = $self->{'jnx'}->get_configuration( compare => "rollback", rollback => "0", format => "text" );
+    my $configcompare = $self->{'jnx'}->get_configuration(
+        compare => "rollback",
+        rollback => "0",
+        format => "text"
+    );
     if ($self->{'jnx'}->has_error) {
-        my $error = $self->{'jnx'}->get_first_error();
+        my $dom = $self->{'jnx'}->get_dom();
+        $self->{'logger'}->error($dom->toString());
 
+        my $error = $self->{'jnx'}->get_first_error();
         if ($error->{'error_message'} !~ /uncommitted/) {
-            $self->set_error($error->{'error_message'});
-
-            $self->{'logger'}->error("Error getting diff from MX: " . $error->{'error_message'});
             $res = $self->{'jnx'}->discard_changes();
-
             $ok = $self->unlock();
-            return;
+            return (undef, "Error getting device diff: $error->{error_message}");
```

Comment:
why not return {status => "success", diff => "blah", error => $error} instead of array?